torch/distributed/fsdp/symbolic_trace.py

```diff
@@ -0,0 +1,202 @@
+                execution_info._traced_param_set.add(p)
+    return proxy
+
+
+def _patched_call_module(
+    call_module: Callable,
+    execution_info: _ExecutionInfo,
+    module: torch.nn.Module,
+    forward: Callable[..., Any],
+    args: Tuple[Any, ...],
+    kwargs: Dict[str, Any]
+) -> Any:
+    """
+    Override of Tracer.call_module (see
+    https://pytorch.org/docs/stable/fx.html#torch.fx.Tracer.call_module).
+    Tracer.call_module is called in symbolic tracing for each non-root module.
+    This override intercepts the recording of each operation and
+    update execution_info.module_forward_order and execution_info.module_execution_info_dict.
+
+    Args:
+        call_module (Callable):
+            The call_module function to be patched.
+        execution_info (_ExecutionInfo):
+            Used to repord the execution information.
+        module, forward, args, kwargs: inputs to the call_module function.
+    """
+    execution_info.module_forward_order.append(module)
+    named_params_list = list(module.named_parameters())
+    if named_params_list != []:
+        execution_info.module_execution_info_dict[execution_info.current_module].append(
+            (module, list(module.named_parameters()))
+        )
+    # Stores away current_module for restoration later
+    old_current_module = execution_info.current_module
+    execution_info.current_module = module
+    # Initialize execution_info.module_execution_info_dict[module]. Note that if
+    # the forward of module is called multiple times, this will record
+    # the execution info of the last forward pass.
+    execution_info.module_execution_info_dict[module] = []
+    # Delegates into the normal Tracer.call_module method
+    output = call_module(module, forward, args, kwargs)
+    # Restores current_module
+    execution_info.current_module = old_current_module
+    return output
+
+
+def _patch_tracer(
+    tracer: torch.fx.Tracer,
+    root_module: torch.nn.Module
+) -> _ExecutionInfo:
+    """
+    Patches the input tracer so that during tracer.trace(), the forward order
+    of all modules and the parameter execution information are recorded.
+    root_module is the top-level module to be traced and should not contain
+    any FSDP modules.
+    """
+    from .fully_sharded_data_parallel import FullyShardedDataParallel
+    for module in root_module.modules():
+        assert (
+            not isinstance(module, FullyShardedDataParallel)
+        ), "The input root_module of _patch_tracer should not contain FSDP modules"
+    execution_info = _ExecutionInfo(root_module)
+    tracer.call_module = functools.partial(
+        _patched_call_module,
+        tracer.call_module,
+        execution_info
+    )
+    params_dict = dict(root_module.named_parameters())
+    tracer.create_proxy = functools.partial(
+        _patched_create_proxy,
+        tracer.create_proxy,
+        execution_info,
+        params_dict
+    )
```

Comment:
Hmm, this looks hacky. It seems the doc suggests creating a custom tracer instead of patch the method on the existing tracer? 

https://pytorch.org/docs/stable/fx.html#customizing-tracing-with-the-tracer-class

cc @jamesr66a

Reply:
@mrshenli yes, this is the part I'm open to discussion. The reason I choose patching is that I hope this would work for all tracers that are not only the instance of `nn.Tracer`, but also the instance of some class that is customized based on `nn.Tracer`. For example, for transformer models developed by hugginface, users may want to use HFTracer.

I have thought about creating a custom tracer class like
```python
class ExecOrderTracer(InputTracer)
```
where InputTracer is the tracer class users want to set as input, but this kind of usage is not common to me and I haven't thought of other good ways to implement this.

Any suggestions would be appreciated!